### ISSUE TITLE
Hide symbols back

### ISSUE THREAD

**Issue by LeonidVas** [tarantool/tarantool#5932]
After #2971 we experience a bunch of problems ([1][1], [2][2]). The second one still have no good solution. You can find more on the topic [here](https://lists.tarantool.org/pipermail/tarantool-discussions/2020-September/000095.html).

Another side of the problem is that I heard from time to time: hey, someone can check that a symbol is present and use it in a module despite that there is no statement that we'll keep it. Yet another side of the problem is that C API description on the website significantly behind the actual state (at least I see no actions around [new APIs](https://github.com/tarantool/doc/issues/1611)), but it is out of scope here.

My proposal:

* Set default visibility to 'hidden' back.
* Either expose the entire public API of a library or hide it and create wrappers / aliases for calls using LuaJIT's FFI or from a Lua/C module.
  * The former (expose) will allow to use the bundled library from a module. It'll disallow to bundle your own library version to the module (tarantool's one will anyway in use on Linux) — say, when you want to use a feature from a newer version of the library. It'll lead to more problems with bundling if the library does not provide stable ABI and intended to be statically linked into a project (like small).
  * The latter (hidding) will disallow to use the bundled library from a module, but will allow to bundle the library into a module or use a system one. We can expose wrappers for necessary APIs if there is a demand (like we [did](https://github.com/tarantool/tarantool/commit/c61179092c08ab1c3e693d4e17b479e119d1b07b) for the region related APIs).

The only advantage of the unhiding symbols: sometimes you can hack something for *certain* tarantool executable using LuaJIT's FFI, but you'll anyway unable to use this hack in a module, because the symbol may disappear in any build: even from the same tarantool version range, but with other OS or toolchain.

The situation for tarantool executable is the same as for any other library. A library should expose only its public API and should not increase probability of hard to debug problems due to clash of a user's code with an internal name from the library.

[1]: https://github.com/tarantool/tarantool/issues/5001
[2]: https://github.com/tarantool/memcached/issues/59

**Comment by LeonidVas** [comment]
See discussion https://github.com/tarantool/tarantool/discussions/5733
